### DISCUSSION TITLE
Notices, warnings, PHP versions and webERP

### DISCUSSION THREAD

**Discussion by pakricard** [timschofield/webERP#235] [Development Discussion & Specification]
Hi all: 

As I keep updating my scripts, I find webERP standard scripts that show Deprecated, Notices or Warnings, depending on the PHP version used.

As a simple example, using the StockStatus.php from v 4.15.2 distribution:

- On PHP 5.6.40 or PHP 7.3.33 it works with no notices or warnings.
- On PHP 7.4.33, it returns eight Notice: Trying to access array offset on value of type null in weberp/StockStatus.php on line XXX
- notices.
- On PHP 8.0.30 or PHP 8.1.30 or PHP 8.2.25 the Notice changes into a Warning

So I have some questions:

- Does it make any sense to develop having error_reporting set to -1?
- Are we aiming to achieve zero notices, or zero warnings in webERP? 
- As it is PHP version dependent and we are getting ready for webERP v5, what PHP version are you considering for it?

Thanks,
Ricard

**Comment by timschofield** [comment]
I always develop using error_reporting set to -1. Though not for
production sites. I know others in the past haven't done so, but I
always think the warnings and notices are there for a reason, and the
PHP developers tend to move them from notices to warnings to errors. I
think it's better to remove them now.

The aim is to get to zero, as I did back in the days of PHP7 but it is
a moving target as the PHP developers create new notices all the time.

My policy is to develop for PHP8 (I use 8.3), but not to use any
functionality that is new in PHP8. This way we should still have
backwards compatibility but most hosting companies seem to have
already moved that way.

Thanks
Tim


On Thu, 14 Nov 2024 at 08:07, pakricard ***@***.***> wrote:
>
> Hi all:
>
> As I keep updating my scripts, I find webERP standard scripts that show Deprecated, Notices or Warnings, depending on the PHP version used.
>
> As a simple example, using the StockStatus.php from v 4.15.2 distribution:
>
> On PHP 5.6.40 or PHP 7.3.33 it works with no notices or warnings.
> On PHP 7.4.33, it returns eight Notice: Trying to access array offset on value of type null in weberp/StockStatus.php on line XXX
> notices.
> On PHP 8.0.30 or PHP 8.1.30 or PHP 8.2.25 the Notice changes into a Warning
>
> So I have some questions:
>
> Does it make any sense to develop having error_reporting set to -1?
> Are we aiming to achieve zero notices, or zero warnings in webERP?
> As it is PHP version dependent and we are getting ready for webERP v5, what PHP version are you considering for it?
>
> Thanks,
> Ricard
>
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you are subscribed to this thread.Message ID: ***@***.***>



-- 
www.weberp.org
@TimSchofield2
Blog: https://kwamoja.home.blog/

**Reply by dalers** [reply]
The PHP version support statement would seem to be "The minimum requirement is PHP 7.2 (7.4?) and is compatible with PHP 8.x up to PHP 8.3." Agree?

**Comment by pakricard** [comment]
Excellent, then Tim. I will continue with error_reporting = -1 and will use PHP8.2.25 (the newest I have available on my server)

Thanks,
Ricard

**Comment by dalers** [comment]
@pakricard what type of server are you using and what OS?

Imho, we need to target PHP 8.3 with no errors or we're not getting ahead of the ball, although I personally use PHP 8.1.30 and will probably wait until 8.1 security support ends before upgrading to give all the web apps I host as much time as possible to be compatible. Other versions available are currently 8.2.24 , 8.3.12 and 8.4.0-rc1 (8.4.0 is to be released next week, and will presumably have 2 years active support followed by 1 year security support)

My suggestion is to officially support v5 only going forwards with latest PHP 8.1 minimum and latest 8.3 recommended - with no PHP or SQL errors or warnings except PHP 8.3 Supporting anything else would be "best available effort".  

For comparison, Drupal officially started supporting 8.3 as of Drupal 10.1 released June 2023 (11.0.7 was released last week). Both Drupal 10.x and 11.x require at least PHP 8.1.6. For WordPress the bar is lower, the current WordPress 6.7 requires PHP 7.2 minimum, but I didn't think v7 was an option for webERP because of the libraries that had to be replaced - and I expect it would help user confidence if the minimum PHP version still had security support. WordPress 6.7 (released this week) is "compatible with exceptions" with PHP 8.0 to 8.2 and "beta" for 8.3. WordPress 6.3 release Aug 2023 was the first release "compatible with exceptions" for PHP 8.1 (fwiw, I have no noticable user or admin issues with WordPress 6.7 on PHP 8.1 using a bundled theme and mimium plugins).

This discussion does bring up the question of how to test and with what versions of PHP and MariaDb/MySQL? Has anyone thought of or tried to create a basic regression test script?

**Comment by pakricard** [comment]
I'm running webERP on a VPS CentOS Linux 7.9.2009 (Core) with Plesk Obsidian 18.0.65 and the newest version of PHP I have available now is 8.2.25. 
DB is MariaDB 5.5.68.
The supplier of the VPS keeps updating versions from time to time, so I will keep updating my development PHP version whenever I have it available.

How to test? In my case, old school ;-) After some development testing, I have 120 users running webERP daily, so I keep pushing changes slowly and if anything is off, I get a report pretty soon. It would be great to have a tool to speed-up testing.

Moreover, I am working with NotePad++to edit PHP and Filezilla to send the files to a development server space (cloned environment from production), and keep debugging with prnMsg(); which is pretty time-consuming. Is there any simple IDE that could accelerate debugging and development that you use? 
P.S.: GitHub just reminded me "This branch is 1007 commits behind timschofield/webERP:master."

**Reply by dalers** [reply]
My server is a discarded office desktop tower in the basement of my house and my only webERP user is me, so you're running a much more sophisticated operation than I am!

I'm hosting a variety of web apps for enginering project management, collaboration, PLM and manufacturing/finance - twelve AMP-type (Apache 2.4, MariaDb 10.6.18, PHP 8.1) and one Python with a PostgreSQL db, but I'm a user first, sysadmin second and developer last.

I'm not aware of any magic to reduce time on development testing, especially given webERP's flat architecture. You can single-step code, set breakpoints and inspect variables in IDEs such as Webstorm and NetBeans, but they take time to learn, configure and use regularly to stay proficient. I've used NetBeans with remote project files before, you don't get single-stepping and integrated debugging, but still benefit from easily searching in all files at once and jumping between files from search results. You might also consider using WinSCP to browse the files on your dev server and launch Notepad++ from within WinSCP for editing (WinSCP copies remote files to local temp files, and then automatically copies them back when you save changes), which at least saves the manual Filezilla step.

I was thinking more of automated regression testing, perhaps using a known database and the API to perform operations and verify results. E.g. something you could run on a release candidate (and particular PHP release) and get high confidence nothing was accidentally broken through unintentional side effects of a believed unrelated change.

I've noticed GitHub added comments to my recent issues #234 #236 and #237 saying you had commits that referenced them. However I don't see how the commits reference the issues. Can you explain what's going on?

Cheers,
Dale

**Reply by dalers** [reply]
P.S. I'm starting to get back into NetBeans and as of now would definately recommend vs Notepad++ and FileZilla. I created a new project with a remote connection to the source code on my server (cloned from my GitHub fork of Tim's "the master" project code repo). NetBeans downloads all files locally and keeps them in sync as files are updated. Iirc, NetBeans includes a git interface but at least for now I'm going to use git on the server in a terminal window (to create a development branch and commit as needed), and use NetBeans only for editing. I'll use GitHub to send a pull requests from my fork to Tim.

I'm going to create a blog post with screenshots when I've got the complete process figured out (happy to donate for developer documentation on weberp.org assuming someone else would find it useful).

**Comment by pakricard** [comment]
Hi @dalers: 
I used to be an IT engineer... last century! But now IT technical side is just a side responsibility. More of a webERP user than a developer. The nice thing about webERP is that we modified it to run exactly as we want the company to operate. Whenever we need to change any process inside the company, we change webERP behaviour accordingly. webERP allows us to run a quite lean operation, and I'm the one maintaining it. 

I haven't committed anything since 2017. In fact I am having a hard time to setup my new sourcetree...

But I see these [pakricard](https://github.com/pakricard) pushed a commit to pakricard/webERP messages in https://github.com/timschofield/webERP/issues/237

Really I have no clue. This morning I tried to setup my SourceTree to read my repository pakricard/webERP and synched it with Tim's. 

Did I mess something up?

**Reply by dalers** [reply]
I don’t think you messed anything up, but I don't understand what happened either. The auto comment was "pakricard pushed a commit to pakricard/webERP that referenced this issue [9 hours ago]". Would you have intentionally pushed to your repo? Perhaps the push was somehow related to your synchronization and a "#xxx" coincidentally appeared in the commit comment?

@timschofield do you understand what is happening?

**Reply by dalers** [reply]
P.S. how (and why) are you synching with Tim's repo? 

What are you doing different compared to the procedure on the webERP wiki? https://www.weberp.org/Development.html

**Comment by timschofield** [comment]
I'm not sure either. You are both forked from my repository so that should
be OK. I had an issue with Phil's repository so in the end I stopped
sending pull requests, as every time I issued a pull request to him, I
appeared to get a mirrored pull request back.

I thought the problem was with his repository but it could be mine. I think
I maybe need to firebase, but I don't want to rush at that.

Tim

On Fri, 15 Nov 2024 at 09:51, Dale Scott ***@***.***> wrote:

> I don’t think you messed anything up, but I don't understand what happened
> either. The auto comment was "pakricard pushed a commit to pakricard/webERP
> that referenced this issue [9 hours ago]". Would you have intentionally
> pushed to your repo? Perhaps the push was somehow related to your
> synchronization and a "#xxx" coincidentally appeared in the commit comment?
>
> @timschofield <https://github.com/timschofield> do you understand what is
> happening?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/timschofield/webERP/discussions/235#discussioncomment-11264961>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAL6LBCZ3QMM2L62SYJJMT2AW7ZBAVCNFSM6AAAAABRYI33SWVHI2DSMVQWIX3LMV43URDJONRXK43TNFXW4Q3PNVWWK3TUHMYTCMRWGQ4TMMI>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
www.weberp.org <http://www.weberpafrica.com>
@TimSchofield2
Blog: https://kwamoja.home.blog/

**Reply by dalers** [reply]
I don't think this is related to the repository.

@timschofield I suspect you made a commit that referenced an issue you created that was actually a list of issues copied from from a Discussion started by @efeoili (btw, good practise to link a commit to an issue it solves! :-)). Then seperate issues were created for each issue on the list, and the list in the issue was updated with references to the individual issues (including three issues I had created). It then seems you may have asssigned or recognised @pakricard as a co-author of your commit, which could be why an auto-comment was made to each of my referenced issues saying pakricard had made a commit that referenced it.

I assume the commit in pakricard's repo was not a direct commit but was from pulling/merging the commit from your repo into his, and GitHub created an auto-comment referencing packricard's repo instead of yours for some reason. I also don't understand the relationship between #212 and #213 but wonder if there's something circular happening. Also I wonder if packricard's "sync" procedure might somehow be involved....

![image](https://github.com/user-attachments/assets/369c4204-fa0b-4c1e-ae81-2432ae9de61f)



![image](https://github.com/user-attachments/assets/61c38fd7-0ead-4803-a975-a58b9798f894)

**Comment by pakricard** [comment]
Yesterday I followed https://www.weberp.org/Development.html to setup my SourceTree, succesfully.

My forked directory (created years ago) was 1.007 commits behind Tim's, and I clicked on "Sync Fork" on my GitHub page.
![Captura de pantalla 2024-11-16 071235](https://github.com/user-attachments/assets/f41c1683-299b-41e8-896b-d13cfa7691b1)

As now, I am one commit behind, I'm going to "sync fork" again, so we can check if it triggers something like yesterday's issue. 

![Captura de pantalla 2024-11-16 071552](https://github.com/user-attachments/assets/03e0af50-d22e-4d17-8241-3e11fd1a0ff7)


![Captura de pantalla 2024-11-16 071655](https://github.com/user-attachments/assets/617a8349-ed31-42db-b23c-63f1fd43b07a)

**Comment by timschofield** [comment]
On Tue, 19 Nov 2024 at 20:02, Dale Scott ***@***.***> wrote:
>
> The PHP version support statement would seem to be "The minimum requirement is PHP 7.2 (7.4?) and is compatible with PHP 8.x up to PHP 8.3." Agree?
>

It should work with 5.x as I have tried to not use new functionality,
but just to remove any warnings etc. However I don't think 5.x is in
use anymore.

**Reply by dalers** [reply]
> However I don't think 5.x is in use anymore.

It might still be, but v5 isn't in the Ubuntu or FreeBSD package repositories anymore (although it can be installed from other sources).

To say v5 is "officially" supported imho there should be at least minimal testing on v5 - which would take time away from other activities. Without testing, and so long as using v5+ language features only doesn't negatively impact webERP development, I suggest claiming webERP _should_ be compatible with v5.x. E.g. "minimum officially supported PHP version is PHP 7.4 but should be compatible with PHP v5.x" (assuming there is testing on v7.4, otherwise "minimum officially supported PHP version is PHP 8.1 but should be compatible with PHP v7.x and v5.x").

At the other end of the spectrum, imho for project credibility webERP needs to follow the PHP release support cycle, i.e. must be compatible with at least v8.1 and preferably 8.3 (8.1 has security support for one more year, 8.2 is just about to end active support and start 2 years of security support and 8.3 has one more year of active support before starting security support).

Even though I'm managing my own bare metal server, it is still problematic for me to use a PHP version that is not actively supported due to the other AMP apps I'm hosting, and I assume I am not the only server administrator in this position (fwiw, the apps I'm hosting are [listed on my website](https://www.dalescott.net/about/saas-applications/), the latest releases of SuiteCRM and Moodle require v8.x, the other apps require or recommend at least 7.4, and none claim support for v5).

**Comment by timschofield** [comment]
I currently have 3 versions on my development machine:

  Selection    Path             Priority   Status
------------------------------------------------------------
  0            /usr/bin/php8.3   83        auto mode
  1            /usr/bin/php5.6   56        manual mode
  2            /usr/bin/php7.2   72        manual mode
* 3            /usr/bin/php8.3   83        manual mode

but 5.6 is just there because I have an old application on there that
I am trying to copy the functionality of.

I am happy to go with your "minimum officially supported PHP version
is PHP 7.4 but should be compatible with PHP v5.x"

Thanks
Tim


On Sat, 23 Nov 2024 at 22:36, Dale Scott ***@***.***> wrote:
>
> However I don't think 5.x is in use anymore.
>
> It might still be, but v5 isn't in the Ubuntu or FreeBSD package repositories anymore (although it can be installed from other sources).
>
> To say v5 is "officially" supported imho there should be at least minimal testing on v5 - which would take time away from other activities. So long as using v5+ language features only doesn't negatively impact webERP development, I suggest claiming webERP should be compatible with v5.x. E.g. "minimum officially supported PHP version is PHP 7.4 but should be compatible with PHP v5.x" (assuming there is testing on v7.4, otherwise "minimum officially supported PHP version is PHP 8.1 but should be compatible with PHP v7.x and v5.x").
>
> At the other end of the spectrum, imho for project credibility webERP needs to follow the PHP release support cycle, i.e. to be compatible with at least v8.1 and preferably 8.3 (8.1 has security support for one more year, 8.2 is just about to start 2 years of security support and 8.3 has one more year of active support before starting security support). Even though I'm managing my own bare metal server, it is still problematic for me to use a PHP version that is not actively supported due to the other AMP apps I'm hosting, and I assume I am not the only server administrator in this position (fwiw, the apps I'm hosting are listed on my website, the latest releases of SuiteCRM and Moodle require v8.x, the other apps require or recommend at least 7.4, and none claim support for v5).
>
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you were mentioned.Message ID: ***@***.***>



-- 
www.weberp.org
@TimSchofield2
Blog: https://kwamoja.home.blog/
